sdio/imx6ull/imx6ull-sdio.c

```diff
@@ -0,0 +1,871 @@
+
+	/* make sure SD clock is stable (SDSTB=0) */
+	do {
+		val = *(sdio_common.base + reg_pres_state) & (1 << 3);
+	} while (!val);
+
+	if (cfg->freqHz == 50000000) {
+		/* FREQ = 400000000 / 2 / (4 * 1) = 50000000 */
+		*(sdio_common.base + reg_sys_ctrl) = (0x2 << 8) | 0xf; /* SDCLKFS=0x2 DVS=0x0 */
+	}
+	else {
+		/* FREQ = 400000000 / 2 / (8 * 1) = 25000000 */
+		*(sdio_common.base + reg_sys_ctrl) = (0x4 << 8) | 0xf; /* SDCLKFS=0x4 DVS=0x0 */
+	}
+
+	/* INITA=1 */
+	val = *(sdio_common.base + reg_sys_ctrl);
+	val |= (1 << 27);
+	*(sdio_common.base + reg_sys_ctrl) = val;
+
+	do {
+		val = *(sdio_common.base + reg_sys_ctrl) & (1 << 27);
+	} while (val);
+
+	return 0;
+}
+
+
+void sdio_free(void)
+{
+	/* reset SDIO and set SDIO clock freq to 400000000 / 2 / (256 * 2) = 390625 Hz */
+	reset_all(0x80, 0x1); /* SDCLKFS=0x80 DVS=0x1 */
+
+	/* REFACTORING NOTE: free resources */
+	if (sdio_common.irq_handle)
+		resourceDestroy(sdio_common.irq_handle);
+	if (sdio_common.eventLock)
+		resourceDestroy(sdio_common.eventLock);
+	if (sdio_common.eventCond)
+		resourceDestroy(sdio_common.eventCond);
+
+	if (sdio_common.cmd_lock)
+		resourceDestroy(sdio_common.cmd_lock);
+
+	if (sdio_common.dmaptr)
+		munmap(sdio_common.dmaptr, DMA_BUFFER_SIZE);
+
+	if (sdio_common.base)
+		munmap((void *)sdio_common.base, _PAGE_SIZE);
+
+	memset(&sdio_common, 0, sizeof(sdio_common));
+}
+
+
+static int sdio_cmdSendWorker(uint8_t cmd, uint32_t arg, uint32_t *res)
+{
+	uint32_t val, cmdFrame;
+
+	if (res)
+		*res = 0;
+
+	val = *(sdio_common.base + reg_pres_state);
+	if (val & 0x7) {
+		printf("libsdio: data interface is busy (pres_state=%x)\n", val);
+		return -1;
+	}
+
+	cmdFrame = (cmd & 0x3f) << 24;
+
+	/* set response type */
+	switch (cmd) {
+		case SDIO_CMD_STATE_IDLE:
+			/* no response */
+			break;
+		case SDIO_CMD_RELATIVE_ADDR:
+			cmdFrame |= 1 << 20;   /* CICEN=1 */
+			cmdFrame |= 1 << 19;   /* CCCEN=1 */
+			cmdFrame |= 0x2 << 16; /* RSPTYP=2 */
+			break;
+		case SDIO_CMD_SEND_OP_COND:
+			cmdFrame |= 0x2 << 16; /* RSPTYP=2 */
+			break;
+		case SDIO_CMD_SELECT_CARD:
+			cmdFrame |= 1 << 20;   /* CICEN=1 */
+			cmdFrame |= 1 << 19;   /* CCCEN=1 */
+			cmdFrame |= 0x3 << 16; /* RSPTYP=3 */
+			break;
+		case SDIO_CMD_STATE_INACTIVE:
+			/* no response */
+			break;
+		case SDIO_CMD_RW_DIRECT:
+			cmdFrame |= 1 << 20;   /* CICEN=1 */
+			cmdFrame |= 1 << 19;   /* CCCEN=1 */
+			cmdFrame |= 0x2 << 16; /* RSPTYP=2 */
+			break;
+		default:
+			return -1;
+	}
+
+	*(sdio_common.base + reg_mix_ctrl) = 1U << 31;
+	*(sdio_common.base + reg_blk_att) = 0;
+	*(sdio_common.base + reg_cmd_arg) = arg;
+	*(sdio_common.base + reg_cmd_xfer_typ) = cmdFrame;
+
+	/* wait 1 ms max */
+	if (wait_for_cmd(0x1, 10, 100) < 0) {
+		reset_cmd_block();
+		return -1;
+	}
+
+	/* clear status flags */
+	*(sdio_common.base + reg_int_status) = 0x1; /* CC=1 */
+
+	/* retrieve response */
+	switch (cmd) {
+		case SDIO_CMD_STATE_IDLE:
+			/* no response */
```

Comment:
Better set `val` to 0 in this case, because now if a caller provides `res` argument you would set it to some random number.